src/lib/optimizer/strategy/data_induced_predicate_rule.cpp

```diff
@@ -0,0 +1,130 @@
+    return side == LQPInputSide::Left ? LQPInputSide::Right : LQPInputSide::Left;
+  };
+
+  const auto estimator = cost_estimator->cardinality_estimator->new_instance();
+  estimator->guarantee_bottom_up_construction();
+
+  visit_lqp(lqp_root, [&estimator, &opposite_side, &data_induced_predicates](const auto& node) {
+    if (node->type != LQPNodeType::Join) {
+      return LQPVisitation::VisitInputs;
+    }
+    const auto join_node = std::static_pointer_cast<JoinNode>(node);
+
+    for (const auto& join_predicate : join_node->join_predicates()) {
+      const auto predicate_expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicate);
+      DebugAssert(predicate_expression, "Expected BinaryPredicateExpression");
+      // Currently only equal predicates are supported.
+      if (predicate_expression->predicate_condition != PredicateCondition::Equals) {
+        continue;
+      }
+
+      // Since data induced predicates might be beneficial for both sides of the join, we create this helper lambda,
+      // which can deal with both sides.
+      const auto reduce_if_beneficial = [&](const auto selection_side) {
+        auto reduced_node = join_node->input(selection_side);
+        auto reducer_node = join_node->input(opposite_side(selection_side));
+
+        auto reducer_side_expression = predicate_expression->left_operand();
+        auto reduced_side_expression = predicate_expression->right_operand();
+
+        auto original_cardinality = estimator->estimate_cardinality(reduced_node);
+
+        /**
+         * We have to check whether the expression is actually evaluable on the reducer_node. If this is not the case
+         * we have to swap the sides. This might be the case, because it is not guaranteed, that the predicates are
+         * ordered according to the input sides.
+         */
+
+        if (!expression_evaluable_on_lqp(reducer_side_expression, *reducer_node)) {
+          std::swap(reduced_side_expression, reducer_side_expression);
+        }
+        DebugAssert(!expression_evaluable_on_lqp(reducer_side_expression, *join_node->input(selection_side)),
+                    "Expected filtered expression to be uniquely evaluable on one side of the join");
+
+        const auto subquery = AggregateNode::make(
+            expression_vector(), expression_vector(min_(reducer_side_expression), max_(reducer_side_expression)),
+            reducer_node);
+
+        const auto min = ProjectionNode::make(expression_vector(min_(reducer_side_expression)), subquery);
+        const auto max = ProjectionNode::make(expression_vector(max_(reducer_side_expression)), subquery);
+
+        // The between predicate is built using the projected min and max from the reducer side expression.
+        auto between_predicate =
+            PredicateNode::make(between_inclusive_(reduced_side_expression, lqp_subquery_(min), lqp_subquery_(max)));
+        lqp_insert_node(join_node, selection_side, between_predicate);
```

Comment:
This would definitely make sense, but because of the limited time, it was not possible for us to implement this. I've added it to the future work! Thanks for spotting that 😊

Reply:
Actually we insert it below the join, but because of the optimizer in the final query plan it is pushed down as far as possible/ the optimizer decides

Reply:
(PredicatePlacementRule and PredicateReorderingRule take care of beneficial placement and are executed after the rewrite.)